Switch from mod_nss to mod_ssl

rcritten opened the pull request:
New installs using an IPA CA should work
Upgrades should work

Not tested and some known to not work:
- [x] CA-less install
- [x] promoting a replica
- [x] promoting a replica CA-less - DL1
- [x] promoting a replica CA-less - DL0
- [x] backup and restore (particularly edge cases like restoring from a mod_nss backup)
- [x] domain level 0 replica install with old master
- [x] domain level 0 replica install with new master

This PR is meant to to obtain status of current patches and as a jumping-off point to finish the rest of the transition.

tiran commented on ipalib/install/certmonger.py:
```diff
@@ -670,11 +670,3 @@ def wait_for_request(request_id, timeout=120):
         raise RuntimeError("request timed out")
 
     return state
-
```
Why did you remove the test block?

rcritten replied:
Because it was broken and generally unused AFAIK so I didn't bother to fix it.

I can fix it if desired, I found it useful during development, but it dated back to the very first commit virtually unchanged (2010).

tiran commented on ipaplatform/base/paths.py:
```diff
@@ -48,6 +48,8 @@ class BasePathNamespace(object):
     HTTPD_IPA_CONF = "/etc/httpd/conf.d/ipa.conf"
     HTTPD_NSS_CONF = "/etc/httpd/conf.d/nss.conf"
     HTTPD_SSL_CONF = "/etc/httpd/conf.d/ssl.conf"
+    HTTPD_CERT_FILE = "/etc/pki/tls/certs/httpd.crt"
```
Are these the default file names for mod_ssl? Although we highly recommend to install IPA on its own machine, people tend to mix installation. I wonder if we should consider or own file names, e.g. ``ipa-httpd``.

rcritten replied:
Sure, it's an easy fix and now is the time to doit.

tiran commented on ipaserver/install/httpinstance.py:
```diff
@@ -237,82 +191,65 @@ def configure_gssproxy(self):
         """This should run after __set_mod_nss_port so is already backed up"""
-        if installutils.update_file(paths.HTTPD_NSS_CONF, '</VirtualHost>', 'Include {path}\n</VirtualHost>'.format(path=paths.HTTPD_IPA_REWRITE_CONF)) != 0:
+        if installutils.update_file(paths.HTTPD_SSL_CONF, '</VirtualHost>', 'Include {path}\n</VirtualHost>'.format(path=paths.HTTPD_IPA_REWRITE_CONF)) != 0:
             print("Adding Include conf.d/ipa-rewrite to %s failed." % paths.HTTPD_NSS_CONF)
```
``HTTPD_SSL_CONF``

rcritten replied:
Gah, this is why we have reviews.

tiran commented on ipaserver/install/httpinstance.py:
```diff
@@ -436,41 +326,50 @@ def __setup_ssl(self):
+                                   paths.HTTPD_KEY_FILE, False)
+        installutils.set_directive(paths.HTTPD_SSL_CONF,
+                                   'SSLCACertificateFile',
+                                   paths.IPA_CA_CRT, False)
```
I don't see support for encrypted key file. Are you planning to add https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslpassphrasedialog later?

rcritten replied:
I think this is something the SPIKE team can add. I was focusing on just getting it working at all.

tiran commented on ipaserver/install/installutils.py:
```diff
@@ -447,10 +447,11 @@ class DirectiveSetter(object):
         ds.set(key, value)
     """
-    def __init__(self, filename, quotes=True, separator=' '):
+    def __init__(self, filename, quotes=True, separator=' ', comment='#'):
```
I like it, but where is ``comment`` argument used? I don't see it.

rcritten replied:
So this requires either further docs or a switch to Augeas.

A lot of the values in the default ssl.conf are commented out which means that new values are appended to the end of the file, outside of the VirtualHost. This can still work but the context is wrong and could really go badly if someone defined other VirtualHost with different trust chains.

I added the comment bit so that a new value that was otherwise commented out could be added near where it should have been. This code is a bit fragile and doesn't account for extra spaces, etc. but it worked in this one small use-case.

Augeas would be more robust.

tiran commented:
Test issues
```
________________ test_set_directive_lines.test_remove_directive ________________
self = <ipatests.test_ipaserver.test_install.test_installutils.test_set_directive_lines object at 0x7f5b86c352e8>
    def test_remove_directive(self):
        lines = installutils.set_directive_lines(
>           False, '=', 'foo', None, EXAMPLE_CONFIG)
E       TypeError: set_directive_lines() missing 1 required positional argument: 'comment'
```

lint issues (try ``make fastcheck``)

```
************* Module ipatests.test_ipaserver.test_install.test_installutils
ipatests/test_ipaserver/test_install/test_installutils.py:32: [E1120(no-value-for-parameter), test_set_directive_lines.test_remove_directive] No value for argument 'comment' in function call)
ipatests/test_ipaserver/test_install/test_installutils.py:37: [E1120(no-value-for-parameter), test_set_directive_lines.test_add_directive] No value for argument 'comment' in function call)
ipatests/test_ipaserver/test_install/test_installutils.py:42: [E1120(no-value-for-parameter), test_set_directive_lines.test_set_directive_does_not_clobber_suffix_key] No value for argument 'comment' in function call)
************* Module ipaserver.install.httpinstance
ipaserver/install/httpinstance.py:25: [W0611(unused-import), ] Unused import pwd)
ipaserver/install/httpinstance.py:26: [W0611(unused-import), ] Unused import re)
ipaserver/install/httpinstance.py:30: [W0611(unused-import), ] Unused import locale)
ipaserver/install/httpinstance.py:33: [W0611(unused-import), ] Unused import six)
ipaserver/install/httpinstance.py:34: [W0611(unused-import), ] Unused Augeas imported from augeas)
ipaserver/install/httpinstance.py:38: [W0611(unused-import), ] Unused EXTERNAL_CA_TRUST_FLAGS imported from ipapython.certdb)
ipaserver/install/httpinstance.py:38: [W0611(unused-import), ] Unused IPA_CA_TRUST_FLAGS imported from ipapython.certdb)
************* Module ipaserver.install.service
ipaserver/install/service.py:399: [W0612(unused-variable), Service.import_ca_certs_file] Unused variable 'ext_key_usage')
ipaserver/install/service.py:399: [W0612(unused-variable), Service.import_ca_certs_file] Unused variable 'nickname')
ipaserver/install/service.py:399: [W0612(unused-variable), Service.import_ca_certs_file] Unused variable 'trusted')
************* Module ipaserver.advise.plugins.smart_card_auth
ipaserver/advise/plugins/smart_card_auth.py:11: [E0611(no-name-in-module), ] No name 'NSS_OCSP_ENABLED' in module 'ipaserver.install.httpinstance')
```

stlaz commented on ipaserver/install/service.py:
```diff
@@ -377,10 +377,40 @@ def add_cert_to_service(self):
+            pass
+        else:
+            with open(cafile, 'wb') as fd:
+                for cert, nickname, trusted, ext_key_usage in ca_certs:
```
Maybe something like `x509.write_certificate_list([cert for cert, _nick, _trust, _eku in ca_certs], cafile)`?

Even if you don't choose that way, you'll probably need to prepend the variable names with "_". I did not check with pylint but I am pretty much sure it will yell at you for unused variables.

rcritten replied:
Fixed as part of the pylint tests fixes. Note that I'm purposely leaving one failing for now, the ipa-advise change, because I'm not sure why OCSP has to be disabled.

stlaz commented on ipaserver/install/service.py:
```diff
@@ -377,10 +377,40 @@ def add_cert_to_service(self):
         if conn is None:
             conn = api.Backend.ldap2
 
+        ca_certs = None
```
`ca_certs` will always be defined in the `else` branch.  Even if the universe goes wrong and it's not , this fix wouldn't make it better because `None` is not an iterable, so the `for` below would still throw an exception.

rcritten replied:
Removed

rcritten commented on ipaserver/install/httpinstance.py:
```diff
@@ -436,41 +326,50 @@ def __setup_ssl(self):
                     certmonger.modify_ca_helper('IPA', prev_helper)
-
-            self.cert = db.get_cert_from_db(self.cert_nickname)
+                self.cert = x509.load_certificate_from_file(
```
Yeah, not sure why I moved it inside, done.

stlaz commented on ipaserver/install/httpinstance.py:
```diff
@@ -345,79 +277,32 @@ def configure_certmonger_renewal_guard(self):
-        self.create_password_conf()
 
         if self.pkcs12_info:
-            if self.ca_is_configured:
```
If you want, I can take care of the CA-less scenario here.

rcritten replied:
That would be awesome and fits my plan to pull other devs in to complete the remaining tasks, thanks!

stlaz commented on ipaserver/install/httpinstance.py:
```diff
@@ -57,52 +53,6 @@
 KDCPROXY_USER = constants.KDCPROXY_USER
 
-# See contrib/nsscipersuite/nssciphersuite.py
-NSS_CIPHER_SUITE = [
```
Should we port this into the format of the `SSLCipherSuite` directive?

tiran replied:
Either we should no longer set a suite at all and leave it to the system policy. Or we should make it a platform decision and don't touch system policies for Fedora and RHEL.

stlaz replied:
I guess it's fine to go with the system policies and possibly add cipher suites in the future in case the defaults are no longer good enough.

rcritten replied:
Right, my goal was to use the system policy.

stlaz commented:
Note to self: this should be tested on some system with FIPS-mode enabled.

stlaz commented:
@rcritten: Are we considering domain level 0 to be supported in an IPA version with this change, too?

rcritten commented:
@stlaz I'm not aware of any reason why DL 0 shouldn't be supported.

stlaz commented:
@rcritten I ask that because you did not handle that case here yet and because it's not yet contained in your TODO list. We'll need to handle HTTPD certs export and import in `ipa-replica-prepare` but that should not be that much of a problem.

rcritten commented:
I think testing DL0 is an excellent thing to add to the TODO list!

stlaz commented:
@rcritten: +1
+TODO: update the `upload_cacrt.py` update plugin.

tiran commented:
linter is still complaining:
```
************* Module ipaserver.install.httpinstance
ipaserver/install/httpinstance.py:31: [W0611(unused-import), ] Unused Augeas imported from augeas)
************* Module ipaserver.advise.plugins.smart_card_auth
ipaserver/advise/plugins/smart_card_auth.py:11: [E0611(no-name-in-module), ] No name 'NSS_OCSP_ENABLED' in module 'ipaserver.install.httpinstance')
```

stlaz commented on ipaserver/install/httpinstance.py:
```diff
@@ -436,41 +321,50 @@ def __setup_ssl(self):
-        ca_db = certs.CertDB(self.realm, nssdir=paths.HTTPD_ALIAS_DIR,
-                             subject_base=self.subject_base)
-        ca_db.export_pem_cert(self.cacert_nickname, paths.CA_CRT)
+        self.import_ca_certs_file(paths.CA_CRT, self.ca_is_configured)
```
Previously, this code would only publish the last CA cert in the chain, now it pastes the whole chain there. This is most probably used only in legacy clients and as a very improbable fallback to receiving the CA cert in `ipa-client-install`.
I wonder - could this behavior change have negative effect in such scenarios? Or should we have actually had the full CA chain there from the beginning?

tiran replied:
See https://pagure.io/freeipa/issue/6454

A CA trust store should only contain trust anchors, not intermediate CA certs. Since most of our infrastructure doesn't support AIA chasing and AIA chasing has also a negative performance impact, intermediate CA certs must be part of the the cert chain in TLS handshake.

rcritten commented:
@tiran I can fix the Augeas issue but I purposely left the failure in the ipa-advise plugin as a reminder that this needs to be fixed properly. I don't know why OCSP has to be disabled in the web server TLS provider for smart cards to work.

rcritten commented:
Turns out I had it totally wrong. The advise script turns **on** OCSP which makes total sense.

I can investigate how to do this in mod_ssl. I think they may have added similar directives.

rcritten commented:
It occurs to me that I changed the upgrade directive for having disabled OCSP. Please give this some thought. It may not really be a problem since the previous update would have reverted in nss.conf which won't exist anymore, or at worst will be reverted to its original backup state.

OCSP is disabled b/c the AVA has ipa-ca built in which can cause issues when IPA DNS is not used.

stlaz commented:
I added 2 patches that should allow us to install freeIPA in CA-less mode, also 2 patches that fix minor code issues. I can create separate PRs for the latter 2 if required.

Please note that the fix of `upload_cacrt.py` is but a hotfix and would probably need to be addressed again in about half a year when 389-ds moves from NSS to SSL, too. I could fix this perhaps more properly by implementing something like `CertBundle` class that would allow us to keep the NSS workflows we grew dependent on, such a class could also encapsulate PEM/PKCS12 files with the actions I am performing on the PKCS#12 cert/keys in `HTTPInstance` in my patches here.

Let me know what you think.

stlaz commented:
Apparently there were some merge conflicts but `git am` did not notify me. I'll see what I can do about it.

rcritten commented:
I had some sort of odd merge patch at HEAD. I dropped it because it was all garbage, rebased against current master and force pushed. Should be saner now.

stlaz commented:
I did the rebase of your branch just yesterday (sorry). Not sure what happened here, but it seems like you took one of Felipe's patches and applied it on top 🤕 
I was looking at the history of the `upload_cacrt.py` file and I guess it's there because we did not always store the CA cert in the database so it's there just for the upgrade path from some very old FreeIPA versions.

stlaz commented:
Did the required changes + fixed a DL0 replica installation issue where we supply an empty password to extract the PKCS#12 file.

stlaz commented:
+TODO: rework `ipa-server-certinstall`

stlaz commented on ipaserver/install/httpinstance.py:
```diff
@@ -237,63 +191,53 @@ def configure_gssproxy(self):
+        # There is no safe way to co-exist since there is no safe port
+        # to make mod_nss use, disable it completely.
+        if os.path.exists(paths.HTTPD_NSS_CONF):
+            installutils.remove_file(paths.HTTPD_NSS_CONF)
```
You can’t just remove a file we don’t own. If the user decides to uninstall IPA, his system may be broken should they try to install something relying on mod_nss, whichever devil that is. Removing the file also breaks the scenario when you uninstall IPA and then try to downgrade the packages and install it again.

rcritten replied:
It should be restored on uninstallation.

The problem is that there is a listen port defined in nss.conf and even if the engine is off Apache is going to listen on that port. So what port should be used? I punted.

What we could do instead is to move nss.conf to nss.conf.ipa in the same directory and create a zero-length nss.conf. If an admin wanted to restore it they could and we would let them deal with the port conflict. It just makes uninstallation quite a bit more complex.

stlaz replied:
My point was exactly as you're saying - it should be restore on uninstallation. You can just push the file to `fstore` that the HTTPInstance uses and restore it at the point where all other files are being restored.

rcritten replied:
I think the scenario I was thinking of was: you have IPA installed with mod_nss, you upgrade and things get converted to mod_ssl and mod_nss is backed up. You uninstall IPA. The restored mod_nss still has IPA configured.

stlaz commented on ipalib/x509.py:
```diff
@@ -535,6 +568,22 @@ def write_certificate_list(certs, filename):
+    """
+    try:
+        with open(filename, 'wb') as fp:
+            fp.write(priv_key.private_bytes(
```
This is potentially dangerous and would write the file as 0x644 if new. I'll be forcing 0x600 in the upcoming update.

stlaz commented:
The latest update fixes the access to written priv. keys + __hopefully__ correctly handles `ipa-certupdate` and `ipa-server-certinstall`

stlaz commented:
While restore succeeds, the WebUI is inaccessible due to a JS error:
```
ErrorcolumnNumber: 377
fileName: "https://masina.address.cool/ipa/ui/js/dojo/dojo.js?v=40690"
info: Array [ "js/freeipa/plugins.js?40690", error ]
lineNumber: 1
message: "scriptError"
src: "dojoLoader"
stack: "l@https://masina.address.cool/ipa/ui/js/dojo/dojo.js?v=40690:1:377\np.injectUrl/o<@https://masina.address.cool/ipa/ui/js/dojo/dojo.js?v=40690:1:12399\n"
__proto__: {…}constructor: function Error()message: ""
name: "Error"
stack: Getter & Setter
toSource: function toSource()
toString: function toString()
__proto__: Object { … }
dojo.js:1:13187
```
I am not a JS developer and have no idea what happens in that minimalized script.

stlaz commented:
The same happens even during a simple ugprade w/o backup-restore, will try to check whether that may be something in master.

pvoborni commented:
Not loading `js/freeipa/plugins.js` and thus breaking Web UI is caused by missing inclusion of redirection rules for https port. I.e. old `nss.conf` has line at the end `Include /etc/httpd/conf.d/ipa-rewrite.conf`, `ssl.conf` doesn't have this line. Adding the config line to `ssl.conf` makes Web UI work again.

stlaz commented:
The upgrade is fixed now. I'll go through the code as I noticed some FIXUP notes I did not address yet before submitting this for another review.

stlaz commented on ipaserver/install/service.py:
```diff
@@ -377,7 +379,36 @@ def add_cert_to_service(self):
                             "%s", self.principal, str(e))
 
-    def import_ca_certs(self, db, ca_is_configured, conn=None):
+    def import_ca_certs_file(self, cafile, ca_is_configured, conn=None):
```
I'm sorry, I just find the name of the method too confusing. Will call this couple `export_ca_certs_{file,nssdb}` in another fixup.

stlaz commented on ipaserver/install/httpinstance.py:
```diff
@@ -436,41 +367,50 @@ def __setup_ssl(self):
-        db = certs.CertDB(self.realm, nssdir=paths.HTTPD_ALIAS_DIR,
-                          subject_base=self.subject_base)
-        self.import_ca_certs(db, self.ca_is_configured)
+        # self.cacert_nickname = db.cacert_name
```
Isn't this done just OK in `__publish_ca_cert`?

stlaz replied:
Never mind, it isn't. We mean to publish the lowest intermediate cert in the chain (at least in external CA case).

stlaz commented:
The latest patchset addresses the FIXME notes and removes some lines that were previously commented-out (that one patch should probably be used in a rebase as fixup).

@rcritten if you're aware of any change I forgot, please, let me know, otherwise this PR could be reviewed.

Once pushed we can think about encrypting the private keys with a password and moving HTTPD_{CERT,KEY}_FILE to an IPA-specifi location (or should the latter be done in this PR?).

stlaz commented:
I added a patch that moves the HTTPD cert/key files to a different location since otherwise a new PR for it would break IPA for any developers working on the master branch and it does not seem to be that much work anyway.

**WARNING:** Rebased on current master, please pull before force-pushing here.

Removing WIP tagging.

rcritten commented on ipaserver/install/httpinstance.py:
```diff
@@ -402,11 +402,14 @@ def __setup_ssl(self):
-            #      raise RuntimeError("Could not find a suitable server cert.")
+            if (priv_key.public_key().public_numbers()
+                    != self.cert.public_key().public_numbers()):
+                raise RuntimeError("Could not find a suitable server cert.")
```
I'd prefer including a more specific error message like "private key does not match public key." IIRC the previous invocation would also verify the cert trust chain was correct but that may not be as easy to do with openssl which is why had the original FIXME. Comparing the keys would be a good first step and there are other restrictions in place elsewhere so this code may be sufficient.

stlaz replied:
Fixed the error message.

I did not find it anywhere in this path but if there was a cert trust chain verification, it must have happened in certmonger while it was adding the certificate to the NSS DB and thus I am assuming it would still be there.

The original `db.find_server_certs()` only returned a list of certificates the database kept private keys of.

rcritten commented on ipaserver/install/httpinstance.py:
```diff
@@ -429,7 +426,14 @@ def configure_mod_ssl_certs(self):
+            raise RuntimeError("HTTPD cert was issued by an unknown CA.")
+        # at this time we can assume any CA cert will be valid since this is
+        # only run during installation
+        x509.write_certificate(ca_certs[0], paths.CA_CRT)
```
Just curious why you went with a list here instead of a single certificate object? If there is more than one matching subject should things blow up anyway?

stlaz replied:
I am not sure there could be a case where there would be multiple certificates with the same subject in `paths.IPA_CA_CRT`. I assumed there would be only one, but the easiest way to get it from list of CA certs in `paths.IPA_CA_CRT` was a list comprehension which still results in a list. If you want, I could perhaps add a check that there's just one cert in the filtered-out list.

tiran replied:
Multiple subjects with the same subject and SKID are a valid use case. For example they are used to provide an alternative chain to an alternative root CA. For example Let's Encrypt is still cross signed by two CAs.

## letsencryptauthorityx3.pem
```
Issuer: C = US, O = Internet Security Research Group, CN = ISRG         
Subject: C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
```

## lets-encrypt-x3-cross-signed.pem
```
Issuer: O = Digital Signature Trust Co., CN = DST Root CA X3
Subject: C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
```

rcritten commented:
The perms on httpd.crt are tighter than they need to be at 0600. Not a deal breaker.

Not sure the scenario matters yet but I installed 4.6.1-3, replaced Apache and 389-ds certs using ipa-server-certinstall, upgraded to master. Things worked ok but uninstall failed because apache couldn't restart. This is because:

httpd: Syntax error on line 353 of /etc/httpd/conf/httpd.conf: Syntax error on line 219 of /etc/httpd/conf.d/ssl.conf: Could not open configuration file /etc/httpd/conf.d/ipa-rewrite.conf: No such file or directory

rcritten commented:
I tested successfully (minus uninstall error mentioned already):

- New install using new bits
- Install on F27 using 4.6.1-3 upgrade to new bits
- Install on F27 CA-less upgrade to new bits
- Install on F27, ipa-server-certinstall, upgrade
- New install on F27, ipa-server-certinstall

I think just the uninstall is a potential blocker.

stlaz commented:
> Not sure the scenario matters yet but I installed 4.6.1-3, replaced Apache and 389-ds certs using ipa-server-certinstall, upgraded to master. Things worked ok but uninstall failed because apache couldn't restart. This is because:

> httpd: Syntax error on line 353 of /etc/httpd/conf/httpd.conf: Syntax error on line 219 of /etc/httpd/conf.d/ssl.conf: Could not open configuration file /etc/httpd/conf.d/ipa-rewrite.conf: No such file or directory

I was not able to reproduce this, could it be that you had a machine that might have possibly had unclean sysrestore or similars? During the uninstallation, HTTPD restores original files like ssl.conf and nss.conf and then tries to restart but the mentioned files get rewritten with their original copies and thus such problems should not occur.

stlaz commented:
Never mind, scratch my previous comment, just reproduced it while doing exactly what you said, previously I thought it was just the upgrade that broke it.
For some reason, the original `ssl.conf` was not backed up in this scenario. That did not cause uninstall failure in my case, but the `ssl.conf` file kept IPA-values. I need to investigate.

stlaz commented:
Found the bug, we actually really weren't storing the original `ssl.conf` in HTTPInstance mod_nss to mod_ssl migration code.

rcritten commented:
I'm seeing the same failure in mod_nss now:

httpd: Syntax error on line 353 of /etc/httpd/conf/httpd.conf: Syntax error on line 216 of /etc/httpd/conf.d/nss.conf: Could not open configuration file /etc/httpd/conf.d/ipa-rewrite.conf: No such file or directory

I think this is because the restored nss.conf is the one from the pre-upgrade state (not the pristine state).

stlaz commented:
Hm, I was counting on the correctness of the docstring in `FileStore.backup_file()` that if a file is already backed up, it won't be again, but apparently it's wrong. I'll see what I can do with this.

stlaz commented:
I added a check that we're not already backing up the `nss.conf` in the 
"httpinstance: backup mod_nss conf instead of just removing it" commit. Untested.

rcritten commented:
I think one final issue, SSLCACertificateFile is getting added to ssl.conf twice. Once after the comment inside the VH and one appended to the end. I think it is because found is not being set.

rcritten commented:
This fixes it and another bug where when a comment match is found a line is lost:

diff --git a/ipaserver/install/installutils.py b/ipaserver/install/installuti
ls.py
index fe481f1c5..885549a80 100644
--- a/ipaserver/install/installutils.py
+++ b/ipaserver/install/installutils.py
@@ -578,8 +578,10 @@ def set_directive_lines(quotes, separator, k, v, li
nes, comment):
             if v is not None:
                 yield new_line
         elif addnext:
+            found = True
             addnext = False
             yield new_line
+            yield line
         elif cmatcher.match(line):
             addnext = True
             yield line

stlaz commented:
Worked for me, was unsure how to push and keep authorship correct so I merged it into `Add value in set_directive after a commented-out version`

stlaz commented:
master:

* 805aea2443c902cd7c5525102701f116b43da575 Use mod_ssl instead of mod_nss for Apache TLS for new installs
* a0407f75f9b25c265008e907c7f97e2e4b0fa6e8 Remove main function from the certmonger library
* 5c64e28512be88f597cec0536576dbaba8b92878 Convert ipa-pki-proxy.conf to use mod_ssl directives
* 4596674481dbbeb9026f0ff793699479e1cfb09d Enable upgrades from a mod_nss-installed master to mod_ssl
* 5531c9f26b1910ec9f852d5ae0da7ddf078ec8fc Don't backup nss.conf on upgrade with the switch to mod_ssl
* 4d2c7a4a75a35d19a1ce069b45d3ac2d5ad80376 Add value in set_directive after a commented-out version
* fa135e6ef123c1ec48463d3389d01363395e2d8e Update smart_card_auth advise script for mod_ssl
* 7dc923cc4c2ec7c01156589f8314c78439dbc857 mod_ssl migration: fix upload_cacrt.py plugin
* 205675239a7ecd66ae77b01e90158b5a559b840e httpinstance: handle supplied PKCS#12 files in installation
* 8789afa18af968032e3f2159a48a6493f6755f47 x509: Remove unused argument of load_certificate_from_file()
* 60aa2c3b38e1f95c465821484f86657fa696e02b x509: Fix docstring of write_certificate()
* c85bf376f06cda17305a8723755f3635a3da275a certupdate: don't update HTTPD NSS db
* 9a7c3159841c10e353861519a1b25bd9b0ef773d Make ipa-server-certinstall store HTTPD cert in a file
* 92d91ed58be62e50aadeac5396602429a02f711e fixup: add ipa-rewrite.conf to ssl.conf on upgrade
* 0c388d1e8f9bdadf83c20ee4186470f0eee593a9 service: rename import_ca_certs_* to export_*
* dde62ff883dbd6f13763f789521d37fca125ea6d httpinstance: backup mod_nss conf instead of just removing it
* 8ea04ab3e3eef431d529c0d99e9d1ea5a65933e1 httpinstance: verify priv key belongs to certificate
* ee49947b6c258a7cff6fa126f06905f297a3e07b httpinstance: fix publishing of CA cert
* 1ca68ea7300daa125eb2fec9aa2de2122b3fd95e httpinstance fixup: remove commented-out lines
* b21941360cdfbcce9fc891befc160103b4678eec Move HTTPD cert/key pair to /var/lib/ipa/certs
* 75845733f86df5ddeba81d8ce5312be0eb511262 Backup ssl.conf when migrating from mod_nss
